ddl/schema.go

```diff
@@ -257,6 +257,124 @@ func onDropSchema(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 	return ver, errors.Trace(err)
 }
 
+func (w *worker) onRecoverSchema(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	var (
+		schemaInfo             *model.DBInfo
+		recoversInfo           []*RecoverInfo
+		dropJobID              int64
+		snapshotTS             uint64
+		oldSchemaName          string
+		recoverSchemaCheckFlag int64
+	)
+	const checkFlagIndexInJobArgs = 4 // The index of `recoverSchemaCheckFlag` in job arg list.
+	if err := job.DecodeArgs(&schemaInfo, &recoversInfo, &dropJobID, &snapshotTS, &recoverSchemaCheckFlag, &oldSchemaName); err != nil {
+		// Invalid arguments, cancel this job.
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	// check GC and safe point
+	gcEnable, err := checkGCEnable(w)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	ver, err = updateSchemaVersion(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	switch schemaInfo.State {
+	case model.StateNone:
+		// none -> write only
+		// check GC enable and update flag.
+		if gcEnable {
+			job.Args[checkFlagIndexInJobArgs] = recoverCheckFlagEnableGC
+		} else {
+			job.Args[checkFlagIndexInJobArgs] = recoverCheckFlagDisableGC
+		}
+		err := checkSchemaNotExists(d, t, schemaInfo.ID, schemaInfo)
+		if err != nil {
+			if infoschema.ErrDatabaseExists.Equal(err) {
+				// The database already exists, can't create it, we should cancel this job now.
+				job.State = model.JobStateCancelled
+			}
+			return ver, errors.Trace(err)
+		}
+		err = t.CreateDatabase(schemaInfo)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		schemaInfo.State = model.StateWriteOnly
+		err = t.UpdateDatabase(schemaInfo)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.SchemaState = model.StateWriteOnly
```

Comment:
Those codes should move into the step `model.StateWriteOnly`(after line 314).

The reason is same as recover table, see the comment in `ddl/table.go(onRecoverTable)`